src/swtpm/mainloop.h

```diff
@@ -54,8 +54,10 @@ struct mainLoopParams {
 #define MAIN_LOOP_FLAG_KEEP_CONNECTION (1 << 2)
 #define MAIN_LOOP_FLAG_READALL    (1 << 3)
 #define MAIN_LOOP_FLAG_END_ON_HUP (1 << 4)
+#define MAIN_LOOP_FLAG_USE_CTRLFD (1 << 5)
 
     int fd;
+    int ctrlfd;
     struct ctrlchannel *cc;
```

Comment:
I think it would be better to hide this in struct ctrlchannel and add a function ctrlchannel_get_client_fd() to retrieve the file descriptor; pass it via --ctrl ...,clientfd=<>. Passing fd= and clientfd= should be mutually exclusive and this can be tested for in ctrlchannel.c . fd is the file descriptor of a listening socket, clientfd the file descriptor of a socket to receive data on directly. Not sure with which type of socket this will actually work.